<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="robot_name" default="MR2"/>
  <arg name="manager_name" default="nodelet_manager"/>
  <arg name="nodelet_mode" default="standalone"/>


  <group ns="$(arg robot_name)">

<!-- Nodelet Manager -->
    <group if="$(eval nodelet_mode=='load')">
      <node pkg="nodelet" type="nodelet" name="$(arg manager_name)" args="manager" />
    </group>

<!-- CAN -->
    <node pkg="nodelet" type="nodelet" name="slcan_bridge" 
    args="$(arg nodelet_mode) can_plugins/SlcanBridge $(arg manager_name)" >
      <param name="port" type="string" value="/dev/ttyACM0" />
    </node>

    <node pkg="nodelet" type="nodelet" name="can_handler" 
    args="$(arg nodelet_mode) can_plugins/CanHandler $(arg manager_name)" />

<!--Motor Driver-->
    <node pkg="nodelet" type="nodelet" name="motor0" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="510" /> 
      <remap from="motor0_cmd" to="foot0_cmd"/>
      <remap from="motor0_cmd_val" to="foot2_cmd_vel"/>      
    </node>
    <node pkg="nodelet" type="nodelet" name="motor1" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="420" /> 
      <remap from="motor1_cmd" to="foot1_cmd"/>
      <remap from="motor1_cmd_val" to="foot1_cmd_vel"/>
    </node>
    <node pkg="nodelet" type="nodelet" name="motor2" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="530" /> 
      <remap from="motor2_cmd" to="foot2_cmd"/>
      <remap from="motor2_cmd_val" to="foot0_cmd_vel"/>
    </node>
    <node pkg="nodelet" type="nodelet" name="motor3" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="540" /> 
      <remap from="motor3_cmd" to="foot3_cmd"/>
      <remap from="motor3_cmd_val" to="foot3_cmd_vel"/>
    </node>
    <!--
    <node pkg="nodelet" type="nodelet" name="motor4" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="410" /> 
      <remap from="motor4_cmd" to="steer2_cmd"/>
      <remap from="motor4_cmd_val" to="steer2_val"/>      
    </node>
    <node pkg="nodelet" type="nodelet" name="motor5" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="420" />
      <remap from="motor5_cmd" to="steer1_cmd"/>
      <remap from="motor5_cmd_val" to="steer1_val"/>
    </node>
    <node pkg="nodelet" type="nodelet" name="motor6" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="430" /> 
      <remap from="motor6_cmd" to="steer0_cmd"/>
      <remap from="motor6_cmd_val" to="steer0_val"/>
    </node>
    <node pkg="nodelet" type="nodelet" name="motor7" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="440" /> 
      <remap from="motor7_cmd" to="steer3_cmd"/>
      <remap from="motor7_cmd_val" to="steer3_val"/>
    </node>
    -->

    <node pkg="nodelet" type="nodelet" name="motor8" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="610" />
      <remap from="motor8_cmd" to="arm_r_cmd"/>
      <remap from="motor8_cmd_val" to="arm_r_value"/>
    </node>
    <node pkg="nodelet" type="nodelet" name="motor9" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="620" />
      <remap from="motor9_cmd" to="arm_l_cmd"/>
      <remap from="motor9_cmd_val" to="arm_l_value"/>
    </node>
    <node pkg="nodelet" type="nodelet" name="motor10" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="630" />
      <remap from="motor10_cmd" to="defence_lift_cmd"/>
      <remap from="motor10_cmd_val" to="defence_lift_value"/>
    </node>
    <node pkg="nodelet" type="nodelet" name="motor11" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="640" />
      <remap from="motor11_cmd" to="defence_roll_cmd"/>
      <remap from="motor11_cmd_val" to="defence_roll_value"/>
    </node>
    <!-- <node pkg="nodelet" type="nodelet" name="motor10" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="52c"/>
      <remap from="motor10_cmd" to="pitch_right_cmd"/>
      <remap from="motor10_cmd_val" to="pitch_right_cmd_pos"/>
    </node>
    <node pkg="nodelet" type="nodelet" name="motor11" args="$(arg nodelet_mode) can_plugins/Md201xNode $(arg manager_name)" >
      <param name="bid" type="string" value="700" />
      <remap from="motor11_cmd" to="pitch_left_cmd"/>
      <remap from="motor11_cmd_val" to="pitch_left_cmd_pos"/>
    </node> -->
    
    
    
    <node pkg="nodelet" type="nodelet" name="MR2_nodelet_main" args="$(arg nodelet_mode) MR2/MR2_nodelet_main $(arg manager_name)" output="screen">
      
      <!--<param name="steer3_adjust_sub" type="double" value="-0.9" />
      <param name="steer0_adjust_sub" type="double" value="0.7" />
      <param name="steer1_adjust_sub" type="double" value="0.9" />
      <param name="steer2_adjust_sub" type="double" value="3.1" />-->
      <param name="defence_lift_upper_pos" type="double" value="40.0" />
      <param name="defence_lift_lower_pos" type="double" value="0.0" />
      <param name="defence_lift_lagoribase_pos" type="double" value="10.0" />
      <param name="defence_roll_horizontal_pos" type="double" value="0.0" />
      <param name="arm_r_avoid1lagori_pos" type="double" value="20.0" />
      <param name="arm_l_avoid1lagori_pos" type="double" value="-20.0" />
      <param name="arm_r_avoidlagoribase_pos" type="double" value="55.0" />
      <param name="arm_l_avoidlagoribase_pos" type="double" value="-55.0" />
    </node>
    
<!--Solenoid_Driver-->  
    <node pkg="solenoid_driver_node" type="solenoid_driver_node" name="solenoid1" >
      <param name="bid" type="string" value="100" />
      <param name="name" value="solenoid1"/>
      <remap from="solenoid1_cmd" to="solenoid1_cmd"/>
      <remap from="solenoid1_order" to="solenoid1_order"/>
    </node>
    <node pkg="solenoid_driver_node" type="solenoid_driver_node" name="solenoid2" >
      <param name="bid" type="string" value="140" />
      <param name="name" value="solenoid2"/>
      <remap from="solenoid2_cmd" to="solenoid2_cmd"/>
      <remap from="solenoid2_order" to="solenoid2_order"/>
    </node>
    <node pkg="solenoid_driver_node" type="solenoid_driver_node" name="solenoid3" >
      <param name="bid" type="string" value="150" />
      <param name="name" value="solenoid3"/>
      <remap from="solenoid3_cmd" to="solenoid3_cmd"/>
      <remap from="solenoid3_order" to="solenoid3_order"/>
    </node>
    <node pkg="solenoid_driver_node" type="solenoid_driver_node" name="solenoid4" >
      <param name="bid" type="string" value="160" />
      <param name="name" value="solenoid4"/>
      <remap from="solenoid4_cmd" to="solenoid4_cmd"/>
      <remap from="solenoid4_order" to="solenoid4_order"/>
    </node>
<!-- joy -->
    <rosparam file="$(find MR2)/param/logitech_f710_joy_assign.yaml" command="load" />
    <node respawn="true" pkg="joy" type="joy_node" name="joy_node" >
      <param name="dev" type="string" value="/dev/input/js0" />
      <param name="deadzone" value="0.12" />
    </node>
    
<!-- key&mouse -->
  <arg name="mouse_vel" default="mouse_vel"/>
    <arg name="holonomic" default="false"/>
    <arg name="plot" default="false"/>
    <node pkg="mouse_teleop" type="mouse_teleop.py" name="mouse_teleop" output="screen">
      <rosparam file="$(find mouse_teleop)/config/mouse_teleop.yaml" command="load"/>
      <param name="scale" value="1.0"/>
      <remap from="mouse_vel" to="$(arg mouse_vel)"/>
    </node>

    <group if="$(arg plot)">
      <node pkg="rqt_plot" type="rqt_plot" name="rqt_plot"
            args="$(arg mouse_vel)/linear/x $(arg mouse_vel)/linear/y $(arg mouse_vel)/angular/z"/>
    </group>

     <node pkg="nodelet" type="nodelet" name="base_controller" args="$(arg nodelet_mode) base_controller_plugins/Omni4 $(arg manager_name)" >
      <param name="motor_max_acc" value="800.0" />
      <param name="motor_max_vel" value="4000.0" />
      <param name="invert_x" value="false" />
      <param name="invert_y" value="false" />
      <param name="invert_z" value="false" />
      <remap from="motor0_cmd_vel" to="foot0_cmd_vel" />
      <remap from="motor1_cmd_vel" to="foot1_cmd_vel" />
      <remap from="motor2_cmd_vel" to="foot2_cmd_vel" />
      <remap from="motor3_cmd_vel" to="foot3_cmd_vel" />
    </node>

    <!--
    <node pkg="nodelet" type="nodelet" name="base_controller" args="$(arg nodelet_mode) base_controller_plugins/Steering $(arg manager_name)" output="screen">
      <param name="tire_max_acc" value="1000.0" />
      <param name="tire_max_vel" value="3000.0" />
      <param name="speed_coeff" value="1500.0" />
      <param name="invert_x" value="false" />
      <param name="invert_y" value="false" />
      <param name="invert_z" value="true" />
      <param name="steer2_adjust_main" value="1.53" />
      <param name="steer1_adjust_main" value="0.85" />
      <param name="steer0_adjust_main" value="0.18" />
      <param name="steer3_adjust_main" value="-0.65" />
      <remap from="tire0_cmd_vel" to="foot0_cmd_vel" />
      <remap from="tire1_cmd_vel" to="foot1_cmd_vel" />
      <remap from="tire2_cmd_vel" to="foot2_cmd_vel" />
      <remap from="tire3_cmd_vel" to="foot3_cmd_vel" />
      <remap from="steer0_cmd_pos" to="steer0_val" />
      <remap from="steer1_cmd_pos" to="steer1_val" />
      <remap from="steer2_cmd_pos" to="steer2_val" />
      <remap from="steer3_cmd_pos" to="steer3_val" />
    </node>
    -->

  </group>
</launch>   
    
